### ISSUE TITLE
Regenerate NASA SMCE AWS account credentials every 60 days

### ISSUE THREAD

**Issue by yuvipanda** [2i2c-org/infrastructure#2434]
## Cluster status

Every 60 days / two months, we need to re-generate the deployer credentials for each nasa AWS account we manage, this issue describes the status of the accounts and how to do it. Update the status below if you have re-generated credentials.

Cluster | Re-generation history
-|-
`nasa-esdis` | April 23 via #3979, Feb 28 via #3747, Jan 5 via #3575, Dec 6 2023 via #3503
`nasa-ghg` | April 23 via #3979, Feb 28 via #3747, Jan 5 via #3575, Dec 12 2023 via #3528, Nov 19 2023 via #3442, Oct 4 2023 via #3220
`nasa-veda` | April 23 via #3979, Feb 28 via #3747, Jan 5 via #3575, Dec 6 2023 via #3506, Nov 19 2023 via #3442, Oct 4 2023 via #3219

Upcoming regeneration: 23 April

## How to re-generate credentials for deployer

https://github.com/2i2c-org/infrastructure/issues/2339 is related, but this is specifically for the continuous deployer key for individual nasa AWS accounts.

Here's the how to on doing this:

1. Authenticate yourself - see https://repost.aws/knowledge-center/authenticate-mfa-cli on how to
2. ```shell
   cd terraform/aws
   ```
3. ```shell
   CLUSTER_NAME=...
   ```
4. ```shell
   # put yourself in the right workspace
   export TF_WORKSPACE=$CLUSTER_NAME
   ```
5. ```shell
   # replace previous credentials with new
   terraform apply -replace=aws_iam_access_key.continuous_deployer -var-file=projects/$CLUSTER_NAME.tfvars
   ```
6. ```shell
   # write new credentials to file, and then encrypt them in place
   terraform output -raw continuous_deployer_creds > ../../config/clusters/$CLUSTER_NAME/enc-deployer-credentials.secret.json
   sops -i -e ../../config/clusters/$CLUSTER_NAME/enc-deployer-credentials.secret.json
   ```
7. ```shell
   # verify function of new credentials
   deployer use-cluster-credentials $CLUSTER_NAME
   ```
8. ```shell
   # add relevant files updated, and then...
   git commit -m "nasa smce clusters: re-generate deployer credentials"
   ```

**Comment by yuvipanda** [comment]
This needs to be done every 60 days :(

**Comment by jmunroe** [comment]
We're coming up on the next cycle for this.  Is it clear who will actually perform the steps listed above for this cycle?

**Comment by pnasrat** [comment]
@jmunroe Id suggest you create a separate issue for the upcoming regeneration linking to this so it can be prioritized/assigned by @damianavila and @2i2c-org/engineering and not get lost in an older issue

**Comment by pnasrat** [comment]
No longer get errors with terraform plan so following instructions

```
Apply complete! Resources: 2 added, 2 changed, 1 destroyed.
```

Deploy to staging with new credentials 

```
git status -s
 M config/clusters/nasa-veda/enc-deployer-credentials.secret.json
...
Release "staging" has been upgraded. Happy Helming!
NAME: staging
LAST DEPLOYED: Thu Jun  1 14:38:13 2023
NAMESPACE: staging
STATUS: deployed
REVISION: 47
TEST SUITE: None

```

**Comment by pnasrat** [comment]
Oops put fixed against the wrong issue

**Comment by yuvipanda** [comment]
This just expired again, and @freitagb is thankfully helping me sort this one out.

**Comment by consideRatio** [comment]
I'm authenticating myself like this:

1. I've setup a `.aws/credentials` file to reference an access key created
   ```
   # about config: https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html
   # about aws cli env vars: https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html
   # about regenerating creds: https://github.com/2i2c-org/infrastructure/issues/2434
   #
   
   [nasa-veda]
   # https://smce-veda.signin.aws.amazon.com/console
   #
   # TMP=$(aws sts get-session-token --profile nasa-ghg --serial-number <...> --token-code <...>)
   # export AWS_ACCESS_KEY_ID=$(echo $TMP | jq -r .Credentials.AccessKeyId) 
   # export AWS_SECRET_ACCESS_KEY=$(echo $TMP | jq -r .Credentials.SecretAccessKey)      
   # export AWS_SESSION_TOKEN=$(echo $TMP | jq -r .Credentials.SessionToken)
   #
   aws_access_key_id=<...>
   aws_secret_access_key=<...>
   ```
2. I use the following commands to acquire useful CLI creds, relying on `aws` and `jq`
   ```
   TMP=$(aws sts get-session-token --profile nasa-ghg --serial-number arn-for-mfa-device --token-code code-from-token)
   export AWS_ACCESS_KEY_ID=$(echo $TMP | jq -r .Credentials.AccessKeyId) 
   export AWS_SECRET_ACCESS_KEY=$(echo $TMP | jq -r .Credentials.SecretAccessKey)      
   export AWS_SESSION_TOKEN=$(echo $TMP | jq -r .Credentials.SessionToken)
   ```

**Comment by sgibson91** [comment]
@consideRatio there is a command in the deployer that will handle MFA of the CLI for these kinds of accounts. Yuvi added it up after I struggled with the bucket setup.

**Comment by consideRatio** [comment]
I was going to run this, to get it off our hands until after christmas, but I ran into unrelated trouble in the terraform state for both `nasa-veda` and `nasa-ghg`, but not for example `openscapes`.

@yuvipanda I recognize this to be a consequence of https://github.com/2i2c-org/infrastructure/pull/3381 not being applied. I'm not sure if its disruptive for workloads accessing the homedirs currently - but luckily nasa-veda and nasa-ghg are without users atm so I'll go for it.

```terraform
  # aws_efs_mount_target.homedirs will be destroyed
  # (because resource uses count or for_each)
  - resource "aws_efs_mount_target" "homedirs" {
      - availability_zone_id   = "usw2-az2" -> null
      - availability_zone_name = "us-west-2a" -> null
      - dns_name               = "fs-09bd18b3ca23eefa2.efs.us-west-2.amazonaws.com" -> null
      - file_system_arn        = "arn:aws:elasticfilesystem:us-west-2:597746869805:file-system/fs-09bd18b3ca23eefa2" -> null
      - file_system_id         = "fs-09bd18b3ca23eefa2" -> null
      - id                     = "fsmt-0c2bce0a2c878da81" -> null
      - ip_address             = "192.168.30.197" -> null
      - mount_target_dns_name  = "us-west-2a.fs-09bd18b3ca23eefa2.efs.us-west-2.amazonaws.com" -> null
      - network_interface_id   = "eni-0624d23adde1f3de3" -> null
      - owner_id               = "597746869805" -> null
      - security_groups        = [
          - "sg-05fcb798fcd453b9c",
        ] -> null
      - subnet_id              = "subnet-02d33168edc138459" -> null
    }

  # aws_efs_mount_target.homedirs["subnet-02d33168edc138459"] will be created
  + resource "aws_efs_mount_target" "homedirs" {
      + availability_zone_id   = (known after apply)
      + availability_zone_name = (known after apply)
      + dns_name               = (known after apply)
      + file_system_arn        = (known after apply)
      + file_system_id         = "fs-09bd18b3ca23eefa2"
      + id                     = (known after apply)
      + ip_address             = (known after apply)
      + mount_target_dns_name  = (known after apply)
      + network_interface_id   = (known after apply)
      + owner_id               = (known after apply)
      + security_groups        = [
          + "sg-05fcb798fcd453b9c",
        ]
      + subnet_id              = "subnet-02d33168edc138459"
    }

  # aws_efs_mount_target.homedirs["subnet-03725b83b8689d3c2"] will be created
  + resource "aws_efs_mount_target" "homedirs" {
      + availability_zone_id   = (known after apply)
      + availability_zone_name = (known after apply)
      + dns_name               = (known after apply)
      + file_system_arn        = (known after apply)
      + file_system_id         = "fs-09bd18b3ca23eefa2"
      + id                     = (known after apply)
      + ip_address             = (known after apply)
      + mount_target_dns_name  = (known after apply)
      + network_interface_id   = (known after apply)
      + owner_id               = (known after apply)
      + security_groups        = [
          + "sg-05fcb798fcd453b9c",
        ]
      + subnet_id              = "subnet-03725b83b8689d3c2"
    }

  # aws_efs_mount_target.homedirs["subnet-08630a475fe02d001"] will be created
  + resource "aws_efs_mount_target" "homedirs" {
      + availability_zone_id   = (known after apply)
      + availability_zone_name = (known after apply)
      + dns_name               = (known after apply)
      + file_system_arn        = (known after apply)
      + file_system_id         = "fs-09bd18b3ca23eefa2"
      + id                     = (known after apply)
      + ip_address             = (known after apply)
      + mount_target_dns_name  = (known after apply)
      + network_interface_id   = (known after apply)
      + owner_id               = (known after apply)
      + security_groups        = [
          + "sg-05fcb798fcd453b9c",
        ]
      + subnet_id              = "subnet-08630a475fe02d001"
    }
```

**Comment by consideRatio** [comment]
Re-generated, see #3442!

**Comment by yuvipanda** [comment]
Thanks for applying that, @consideRatio. Those are not disruptive, and I had missed applying them. Am getting better at getting things to 100% complete than 90%, but this slipped through - sorry about that.

**Comment by sgibson91** [comment]
I'm now having to do this for `nasa-esdis`. I ran into the same issue described by Erik [here](https://github.com/2i2c-org/infrastructure/issues/2434#issuecomment-1817943343) and applied since they were to be non-disruptive. But I got an error message:

```
│ Error: creating EFS Mount Target (fs-0013506a2d5ee70fc): MountTargetConflict: mount target already exists in this AZ
│ {
│   RespMetadata: {
│     StatusCode: 409,
│     RequestID: "71f43368-7e14-4949-9856-60dbd0d56a78"
│   },
│   ErrorCode: "MountTargetConflict",
│   Message_: "mount target already exists in this AZ"
│ }
│ 
│   with aws_efs_mount_target.homedirs["subnet-0614449fa3a84a06e"],
│   on efs.tf line 70, in resource "aws_efs_mount_target" "homedirs":
│   70: resource "aws_efs_mount_target" "homedirs" {
```

**Comment by sgibson91** [comment]
I also generalised the title a bit

**Comment by sgibson91** [comment]
@damianavila It may need to happen now. There is no Access Key associated with the `hub-continuous-deployer` user in the AWS console, and running `deployer deploy nasa-ghg staging` failed with:

```
An error occurred (UnrecognizedClientException) when calling the DescribeCluster operation: The security token included in the request is invalid.
```

**Comment by damianavila** [comment]
>@damianavila It may need to happen now.

@sgibson91, can you take care of this one before EOW?

**Comment by sgibson91** [comment]
Done

**Comment by consideRatio** [comment]
Re-generated all creds again to keep them aligned as part of doing it for esdis where the deployer key had been deleted for some reason.

**Comment by consideRatio** [comment]
@2i2c-org/engineering I'll assign myself to do regular checks of this every ~50 days or so, aiming for between 1-2 weeks notice.

1. re-generate deployer credentials before they expire
2. check if a member of 2i2c has been blocked after getting expired passwords, unblock them and send new credentials via slack
3. write a reminder in `#engineering` to refresh both cloud console passwords and the any CLI credentials they may have setup

**Comment by yuvipanda** [comment]
This is amazing, thank you so much @consideRatio

**Comment by consideRatio** [comment]
Ah, I think the issue is that the instructions didn't replace existing credentials, they only ensured new ones if the previous were gone.

I've updated the instructions to use a flag for terraform apply, making us get what we want done more directly: `-replace=aws_iam_access_key.continuous_deployer`.

I think due to this, my conclusions on needing to re-generate more often than every 60 days was incorrect - we just had to make sure we successfully re-generated credentials even if there were existing non-expired credentials around already.

**Comment by consideRatio** [comment]
I've scheduled myself to re-generate 23 April, in two weeks.

**Comment by consideRatio** [comment]
I've been asked to not do this task by @yuvipanda, but credentials will expire ~now or have already expired I think, so I've unassigned myself and moved this to the top of the refined backlog.

**Comment by yuvipanda** [comment]
Thank you for not doing it, @consideRatio. It should be handled as part of https://github.com/2i2c-org/infrastructure/issues/4114. Those credentials have expired, because ideally we would have gotten to this as part of last refinement. But we didn't, and that's ok. 

I'm going to actually close this task, as we'll instead be generating new tasks each time based on the calendar mentioned there.
